### DISCUSSION TITLE
Applying tracked changes to recover from optimistic concurrency conflict

### DISCUSSION THREAD

**Discussion by danielwinkler** [ravendb/ravendb#13852] [Q&A] [answered]
Hi all,

I'd have a question regarding optimistic concurrency and change tracking.

When we track an object using the session, it is possible to find all changes that have been applied to it using
`DocumentSession.Advanced.WhatChanged();`.

Is it possible to use this information to recover from a concurrency conflict when using optimistic concurrency, i.e.
- Create session `DocumentSession` and load `example-123`
- Change an object `example-123` locally; session tracks changes
- `DocumentSession.SaveChanges()`
- handle exception
- Create a new session `DocumentSession2`
- load latest version of `var recovery = DocumentSession2.Load("example-123")`
- use the tracked changes from `WhatChanged` and apply to `recovery` 
- `DocumentSession2.SaveChanges()`

Is this possible in general, and does ravenDB provide already functionality to perform the mentioned application of tracked changes to an object?
(Maybe it is also possible to transform the changes to PatchRequests but I didn't find an example for that.)

Kind regards,
Daniel

**Comment by ayende** [comment, marked as answer]
You can write the code to do that, but I would strongly suggest you won't.
The problem is that you are trying to solve generically a merge issue that may have business dependencies.

Consider the following scenario:

* You have an `Order` document show `ShippingAddress` is '"Atlanta, GA"`
* In one thread, you modify the `Order` to add a `Glyphosate` (a herbicide) to the `Items`.
* In another thread, you modify the `ShippingAddress` to be `"Barcelona, Spain"`. You get an optimistic concurrency exception here.

Here is the issue, Glyphosate is banned in Spain, so you are not allowed to ship it. When you add the product to the `Item`, you are running the appropriate validation for that. And when you changed the `ShippingAddress` you also validated that all the `Items` in the `Order` are valid.

However, in your recovery model, what do you do? If you just apply the changes and call `SaveChanges`, you're likely to violate the business rules for the system.

**Reply by danielwinkler** [reply]
Thanks @ayende for your feedback; I agree with the problem that you depict once we'd generalize the automatic merge.

The origin of the problem is that denormalizing a document can result in increased responsibilities, thus more actors that work on it.

I saw your webinar about event sourcing which seems to be a very elegant solution to the problem.
Coping with this problem in a CRUD fashion will require the client to resubmit the request; which seems to be the only safe option left.

**Reply by ayende** [reply]
The key issue is what sort of operations are you allowing, yes. Taking the same logic further, the client will not be submitting the same request, they would be consider this again and maybe not doing this.

A good example would be bug tracking, if you and I are trying to mark the same bug as 'in progress', or 'assigned to me', that is not an action that you want to just blindly retry.
